# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\pavilion\\Desktop\\Rahul\\Springboard\\Springboard_May_20\\material\\GuidedCapstoneProject\\Step 2\\data'

In [3]:
os.listdir()

['.ipynb_checkpoints',
 '1588204767_GuidedCapstoneStep3HL.ipynb (4).zip',
 '1588204778_GuidedCapstoneStep4_and_step5HL.ipynb.zip',
 'GuidedCapstoneStep3HL.ipynb',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'NOtebook_Stepfour.ipynb',
 'Notebook_Stepthree.ipynb',
 'step2_output.csv',
 'step3_output.csv',
 '__MACOSX']

In [4]:
df = pd.read_csv('step3_output.csv')

In [5]:
df.head()

Unnamed: 0  Unnamed: 0.1                           Name       state  \
0           2             2               Hilltop Ski Area      Alaska   
1           4             4            Sunrise Park Resort     Arizona   
2           5             5  Yosemite Ski & Snowboard Area  California   
3           8             8         Boreal Mountain Resort  California   
4           9             9                    Dodge Ridge  California   

   summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  ...  \
0         2090            294      0        0.0          0          0  ...   
1        11100           1800      0        0.0          0          1  ...   
2         7800            600      0        0.0          0          0  ...   
3         7700            500      0        0.0          0          1  ...   
4         8200           1600      0        0.0          0          0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        115.000000       49.0   
2               88.0             0.0        110.000000       84.0   
3              380.0           200.0        150.000000       54.0   
4              862.0             0.0        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         2  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 28 columns]

In [6]:
df.drop(['Unnamed: 0.1'],axis=1,inplace=True)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175 entries, 0 to 174
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         175 non-null    int64  
 1   Name               175 non-null    object 
 2   state              175 non-null    object 
 3   summit_elev        175 non-null    int64  
 4   vertical_drop      175 non-null    int64  
 5   trams              175 non-null    int64  
 6   fastEight          175 non-null    float64
 7   fastSixes          175 non-null    int64  
 8   fastQuads          175 non-null    int64  
 9   quad               175 non-null    int64  
 10  triple             175 non-null    int64  
 11  double             175 non-null    int64  
 12  surface            175 non-null    int64  
 13  total_chairs       175 non-null    int64  
 14  Runs               175 non-null    float64
 15  TerrainParks       175 non-null    float64
 16  LongestRun_mi      175 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [8]:
#dfo=df.select_dtypes(include=['object']) # select object type columns
#df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

In [9]:
df_new = df# copying the original df in df_new

In [10]:
df = df.drop(['Name'],axis=1)
#df[df.columns.difference(['Name'])]

In [11]:
dfo = df.select_dtypes(include=['object'])

In [12]:
dfo.head()

state
0      Alaska
1     Arizona
2  California
3  California
4  California

In [13]:
df = pd.concat([df.drop(dfo,axis=1),pd.get_dummies(dfo)],axis=1)

In [14]:
df.head()

Unnamed: 0  summit_elev  vertical_drop  trams  fastEight  fastSixes  \
0           2         2090            294      0        0.0          0   
1           4        11100           1800      0        0.0          0   
2           5         7800            600      0        0.0          0   
3           8         7700            500      0        0.0          0   
4           9         8200           1600      0        0.0          0   

   fastQuads  quad  triple  double  ...  state_Rhode Island  \
0          0     0       1       0  ...                   0   
1          1     2       3       1  ...                   0   
2          0     0       1       3  ...                   0   
3          1     1       3       1  ...                   0   
4          0     1       2       5  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 60 columns]

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175 entries, 0 to 174
Data columns (total 60 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            175 non-null    int64  
 1   summit_elev           175 non-null    int64  
 2   vertical_drop         175 non-null    int64  
 3   trams                 175 non-null    int64  
 4   fastEight             175 non-null    float64
 5   fastSixes             175 non-null    int64  
 6   fastQuads             175 non-null    int64  
 7   quad                  175 non-null    int64  
 8   triple                175 non-null    int64  
 9   double                175 non-null    int64  
 10  surface               175 non-null    int64  
 11  total_chairs          175 non-null    int64  
 12  Runs                  175 non-null    float64
 13  TerrainParks          175 non-null    float64
 14  LongestRun_mi         175 non-null    float64
 15  SkiableTerrain_ac     1

In [16]:
pd.get_dummies(dfo)

state_Alaska  state_Arizona  state_California  state_Colorado  \
0               1              0                 0               0   
1               0              1                 0               0   
2               0              0                 1               0   
3               0              0                 1               0   
4               0              0                 1               0   
..            ...            ...               ...             ...   
170             0              0                 0               0   
171             0              0                 0               0   
172             0              0                 0               0   
173             0              0                 0               0   
174             0              0                 0               0   

     state_Connecticut  state_Idaho  state_Illinois  state_Indiana  \
0                    0            0               0              0   
1                    0            0               0              0   
2                    0            0               0              0   
3                    0            0               0              0   
4                    0            0               0              0   
..                 ...          ...             ...            ...   
170                  0            0               0              0   
171                  0            0               0              0   
172                  0            0               0              0   
173                  0            0               0              0   
174                  0            0               0              0   

     state_Iowa  state_Maine  ...  state_Rhode Island  state_South Dakota  \
0             0            0  ...                   0                   0   
1             0            0  ...                   0                   0   
2             0            0  ...                   0                   0   
3             0            0  ...                   0                   0   
4             0            0  ...                   0                   0   
..          ...          ...  ...                 ...                 ...   
170           0            0  ...                   0                   0   
171           0            0  ...                   0                   0   
172           0            0  ...                   0                   0   
173           0            0  ...                   0                   0   
174           0            0  ...                   0                   0   

     state_Tennessee  state_Utah  state_Vermont  state_Virginia  \
0                  0           0              0               0   
1                  0           0              0               0   
2                  0           0              0               0   
3                  0           0              0               0   
4                  0           0              0               0   
..               ...         ...            ...             ...   
170                0           0              0               0   
171                0           0              0               0   
172                0           0              0               0   
173                0           0              0               0   
174                0           0              0               0   

     state_Washington  state_West Virginia  state_Wisconsin  state_Wyoming  
0                   0                    0                0              0  
1                   0                    0                0              0  
2                   0                    0                0              0  
3                   0                    0                0              0  
4                   0                    0                0              0  
..                ...                  ...              ...            ...  
170                 0                    0            

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [17]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [18]:
X.head(2)

Unnamed: 0  summit_elev  vertical_drop  trams  fastEight  fastSixes  \
0           2         2090            294      0        0.0          0   
1           4        11100           1800      0        0.0          0   

   fastQuads  quad  triple  double  ...  state_Rhode Island  \
0          0     0       1       0  ...                   0   
1          1     2       3       1  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  

[2 rows x 59 columns]

In [19]:
y.head(2)

0    34.0
1    78.0
Name: AdultWeekend, dtype: float64

In [20]:
scaler

StandardScaler(copy=True, with_mean=True, with_std=True)

In [21]:
X_scaled

array([[-1.76162569, -0.4996967 , -0.95391337, ..., -0.13206764,
        -0.24618298, -0.17149859],
       [-1.7403368 ,  2.28443017,  1.59817699, ..., -0.13206764,
        -0.24618298, -0.17149859],
       [-1.72969235,  1.26471667, -0.43536115, ..., -0.13206764,
        -0.24618298, -0.17149859],
       ...,
       [ 1.69781983,  1.2671887 ,  1.21011013, ..., -0.13206764,
        -0.24618298,  5.83095189],
       [ 1.70846428,  1.84039129,  0.22553875, ..., -0.13206764,
        -0.24618298,  5.83095189],
       [ 1.71910873,  1.79002362,  0.41194641, ..., -0.13206764,
        -0.24618298,  5.83095189]])

In [22]:
X_scaled.shape

(175, 59)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [23]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [24]:
y[0:5]

array([34.        , 78.        , 47.        , 64.16681004, 78.        ])

In [25]:
y.shape

(175,)

In [26]:
X_train.shape

(131, 59)

In [27]:
for i,col in enumerate(X):
    print(i,col)

0 Unnamed: 0
1 summit_elev
2 vertical_drop
3 trams
4 fastEight
5 fastSixes
6 fastQuads
7 quad
8 triple
9 double
10 surface
11 total_chairs
12 Runs
13 TerrainParks
14 LongestRun_mi
15 SkiableTerrain_ac
16 Snow Making_ac
17 daysOpenLastYear
18 yearsOpen
19 averageSnowfall
20 AdultWeekday
21 projectedDaysOpen
22 NightSkiing_ac
23 clusters
24 state_Alaska
25 state_Arizona
26 state_California
27 state_Colorado
28 state_Connecticut
29 state_Idaho
30 state_Illinois
31 state_Indiana
32 state_Iowa
33 state_Maine
34 state_Maryland
35 state_Massachusetts
36 state_Michigan
37 state_Minnesota
38 state_Missouri
39 state_Montana
40 state_Nevada
41 state_New Hampshire
42 state_New Jersey
43 state_New Mexico
44 state_New York
45 state_North Carolina
46 state_Ohio
47 state_Oregon
48 state_Pennsylvania
49 state_Rhode Island
50 state_South Dakota
51 state_Tennessee
52 state_Utah
53 state_Vermont
54 state_Virginia
55 state_Washington
56 state_West Virginia
57 state_Wisconsin
58 state_Wyoming


In [34]:
X2 = X_train[:,:24]
X2#used in model 2

array([[-0.60138089, -0.80406573, -0.85901493, ..., -0.48331852,
         1.17856766,  0.76809537],
       [-0.88878098, -0.61402821,  0.54751562, ...,  0.37427021,
         0.75488372,  0.76809537],
       [-0.05851405, -0.5451203 ,  0.18995183, ..., -0.34038707,
        -0.93985203,  0.76809537],
       ...,
       [-0.30333635, -0.80561075, -0.94374568, ..., -1.29326344,
        -0.30432612,  0.76809537],
       [ 0.99528628,  1.51191994,  1.25925397, ...,  0.37427021,
        -0.93985203, -0.57607153],
       [-1.0058699 , -0.81827992, -0.64718804, ...,  0.3767955 ,
         0.2252788 ,  0.76809537]])

In [33]:
X.columns[:24]
#used in model 2

Index(['Unnamed: 0', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [32]:
X2.shape

(131, 24)

In [28]:
X3=X.iloc[:,np.r_[0,3:24]]
X3#used in model 3

Unnamed: 0  trams  fastEight  fastSixes  fastQuads  quad  triple  double  \
0             2      0        0.0          0          0     0       1       0   
1             4      0        0.0          0          1     2       3       1   
2             5      0        0.0          0          0     0       1       3   
3             8      0        0.0          0          1     1       3       1   
4             9      0        0.0          0          0     1       2       5   
..          ...    ...        ...        ...        ...   ...     ...     ...   
170         323      0        0.0          0          0     0       0       1   
171         326      0        0.0          0          0     0       1       1   
172         327      0        0.0          0          0     1       1       1   
173         328      0        0.0          0          0     0       1       3   
174         329      0        0.0          0          0     0       2       0   

     surface  total_chairs  ...  LongestRun_mi  SkiableTerrain_ac  \
0          2             3  ...            1.0               30.0   
1          0             7  ...            1.2              800.0   
2          1             5  ...            0.4               88.0   
3          2             8  ...            1.0              380.0   
4          4            12  ...            2.0              862.0   
..       ...           ...  ...            ...                ...   
170        1             2  ...            0.6               92.0   
171        1             3  ...            1.0              184.0   
172        0             3  ...            1.0              400.0   
173        1             5  ...            0.7               75.0   
174        0             2  ...            0.4              370.0   

     Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  \
0              30.0        150.000000       36.0             69.0   
1              80.0        115.000000       49.0            250.0   
2               0.0        110.000000       84.0            300.0   
3             200.0        150.000000       54.0            400.0   
4               0.0        115.103943       69.0            350.0   
..              ...               ...        ...              ...   
170            32.0        121.000000       61.0             80.0   
171            18.0         61.000000       81.0            310.0   
172           250.0        121.000000       80.0            300.0   
173            30.0        131.000000       59.0            250.0   
174             0.0        115.103943       81.0            150.0   

     AdultWeekday  projectedDaysOpen  NightSkiing_ac  clusters  
0       30.000000         152.000000            30.0         2  
1       74.000000         104.000000            80.0         1  
2       47.000000         107.000000             0.0         1  
3       49.000000         150.000000           200.0         1  
4       78.000000         140.000000             0.0         1  
..            ...                ...             ...       ...  
170     48.000000          95.000000             0.0         1  
171     42.000000          77.000000             0.0         1  
172     59.000000         123.000000           110.0         1  
173     49.000000         120.053004             0.0         1  
174     57.916957         120.053004             0.0         1  

[175 rows x 22 columns]

In [29]:
X3.columns

Index(['Unnamed: 0', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad',
       'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [30]:
X3.shape

(175, 22)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [35]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [36]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [37]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [38]:
from sklearn.metrics import explained_variance_score,mean_absolute_error
evs = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)

In [39]:
evs

-2.7434902015412724e+21

In [40]:
mae

109791169954.2541

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [41]:
lm.intercept_

27604637071.571003

In [42]:
lm.coef_

array([ 1.72097691e+01, -1.05653819e+00,  1.05974075e+00, -1.46047269e+13,
       -1.79634285e+13,  6.64074510e+12,  8.27584315e+11,  9.89020378e+11,
        1.73512003e+12,  2.11653309e+12,  2.02973526e+12, -3.87378118e+12,
       -2.87109375e-01,  1.91552734e+00,  1.38024902e+00,  1.61132812e-02,
        7.55859375e-01,  1.43603516e+00, -4.25292969e-01,  1.06250000e+00,
        7.65332031e+00, -2.07080078e+00,  2.79296875e-01, -6.24023438e-01,
       -3.64129938e+11, -3.64129938e+11, -1.22086091e+12, -8.04806037e+11,
       -8.04806037e+11, -8.79024015e+11, -6.27056413e+11, -5.13475601e+11,
       -6.27056413e+11, -8.04806037e+11, -3.64129938e+11, -7.21954477e+11,
       -1.35234464e+12, -1.00898566e+12, -5.13475601e+11, -7.21954477e+11,
       -3.64129938e+11, -1.17246538e+12,  2.33126251e+05, -8.79024015e+11,
       -1.53696016e+12, -6.27056413e+11, -7.21954477e+11, -6.27056413e+11,
       -1.17246538e+12, -3.64129938e+11, -3.64129938e+11, -3.64129938e+11,
       -6.27056413e+11, -

In [43]:
lm.coef_.shape

(59,)

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [44]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df_coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df_coef

Coefficient
Unnamed: 0            1.720977e+01
summit_elev           1.056538e+00
vertical_drop         1.059741e+00
trams                 1.460473e+13
fastEight             1.796343e+13
fastSixes             6.640745e+12
fastQuads             8.275843e+11
quad                  9.890204e+11
triple                1.735120e+12
double                2.116533e+12
surface               2.029735e+12
total_chairs          3.873781e+12
Runs                  2.871094e-01
TerrainParks          1.915527e+00
LongestRun_mi         1.380249e+00
SkiableTerrain_ac     1.611328e-02
Snow Making_ac        7.558594e-01
daysOpenLastYear      1.436035e+00
yearsOpen             4.252930e-01
averageSnowfall       1.062500e+00
AdultWeekday          7.653320e+00
projectedDaysOpen     2.070801e+00
NightSkiing_ac        2.792969e-01
clusters              6.240234e-01
state_Alaska          3.641299e+11
state_Arizona         3.641299e+11
state_California      1.220861e+12
state_Colorado        8.048060e+11
state_Connecticut     8.048060e+11
state_Idaho           8.790240e+11
state_Illinois        6.270564e+11
state_Indiana         5.134756e+11
state_Iowa            6.270564e+11
state_Maine           8.048060e+11
state_Maryland        3.641299e+11
state_Massachusetts   7.219545e+11
state_Michigan        1.352345e+12
state_Minnesota       1.008986e+12
state_Missouri        5.134756e+11
state_Montana         7.219545e+11
state_Nevada          3.641299e+11
state_New Hampshire   1.172465e+12
state_New Jersey      2.331263e+05
state_New Mexico      8.790240e+11
state_New York        1.536960e+12
state_North Carolina  6.270564e+11
state_Ohio            7.219545e+11
state_Oregon          6.270564e+11
state_Pennsylvania    1.172465e+12
state_Rhode Island    3.641299e+11
state_South Dakota    3.641299e+11
state_Tennessee       3.641299e+11
state_Utah            6.270564e+11
state_Vermont         1.008986e+12
state_Virginia        5.134756e+11
state_Washington      7.219545e+11
state_West Virginia   6.270564e+11
state_Wisconsin       1.121305e+12
state_Wyoming         8.048060e+11

In [45]:
df_coef.sort_values(by=['Coefficient'],ascending=False).head(11)

Coefficient
fastEight           1.796343e+13
trams               1.460473e+13
fastSixes           6.640745e+12
total_chairs        3.873781e+12
double              2.116533e+12
surface             2.029735e+12
triple              1.735120e+12
state_New York      1.536960e+12
state_Michigan      1.352345e+12
state_California    1.220861e+12
state_Pennsylvania  1.172465e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

In [46]:
#list(X_train.filter(regex='State'))

#### Model 2

In [47]:
X2_train = X_train[:,:24]
X2_test = X_test[:,:24]
y2_train = y_train
y2_test = y_test

In [48]:
X2_train.shape

(131, 24)

In [49]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [50]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred2 = model2.predict(X2_test)

In [51]:
from sklearn.metrics import explained_variance_score,mean_absolute_error
evs2 = explained_variance_score(y2_test,y_pred2)
mae2 = mean_absolute_error(y2_test,y_pred2)

In [52]:
evs2

0.6540519043831821

In [53]:
mae2

6.923908686959541

In [54]:
lm2.intercept_

56.34184694503338

In [55]:
lm2.coef_

array([ 3.76928176e-01, -3.73979867e+00,  2.08178087e+00, -1.55431223e-15,
       -1.55431223e-15, -1.11022302e-15,  5.58171364e-01,  1.66626574e+00,
        6.50650074e-01,  8.13519617e-01, -1.45657538e+00,  5.17385804e-01,
        1.69538083e-01,  1.90964509e+00,  1.01226518e+00, -5.70879247e-01,
        1.06579767e+00,  2.06934889e+00, -1.66383186e-01, -2.18289407e-01,
        9.73525462e+00, -2.02172624e+00, -3.63333806e-01, -2.75382194e+00])

In [56]:
lm2.coef_.shape

(24,)

In [57]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df_coef2 = pd.DataFrame(abs(lm2.coef_), X.columns[:24], columns=['Coefficient'])
df_coef2

Coefficient
Unnamed: 0         3.769282e-01
summit_elev        3.739799e+00
vertical_drop      2.081781e+00
trams              1.554312e-15
fastEight          1.554312e-15
fastSixes          1.110223e-15
fastQuads          5.581714e-01
quad               1.666266e+00
triple             6.506501e-01
double             8.135196e-01
surface            1.456575e+00
total_chairs       5.173858e-01
Runs               1.695381e-01
TerrainParks       1.909645e+00
LongestRun_mi      1.012265e+00
SkiableTerrain_ac  5.708792e-01
Snow Making_ac     1.065798e+00
daysOpenLastYear   2.069349e+00
yearsOpen          1.663832e-01
averageSnowfall    2.182894e-01
AdultWeekday       9.735255e+00
projectedDaysOpen  2.021726e+00
NightSkiing_ac     3.633338e-01
clusters           2.753822e+00

In [58]:
df_coef2.sort_values(by=['Coefficient'],ascending=False).head(11)

Coefficient
AdultWeekday          9.735255
summit_elev           3.739799
clusters              2.753822
vertical_drop         2.081781
daysOpenLastYear      2.069349
projectedDaysOpen     2.021726
TerrainParks          1.909645
quad                  1.666266
surface               1.456575
Snow Making_ac        1.065798
LongestRun_mi         1.012265

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [59]:
X2_train.shape

(131, 24)

In [60]:
np.delete(X2_train,[1,2],1)

array([[-0.60138089,  0.        ,  0.        , ..., -0.48331852,
         1.17856766,  0.76809537],
       [-0.88878098,  0.        ,  0.        , ...,  0.37427021,
         0.75488372,  0.76809537],
       [-0.05851405,  0.        ,  0.        , ..., -0.34038707,
        -0.93985203,  0.76809537],
       ...,
       [-0.30333635,  0.        ,  0.        , ..., -1.29326344,
        -0.30432612,  0.76809537],
       [ 0.99528628,  0.        ,  0.        , ...,  0.37427021,
        -0.93985203, -0.57607153],
       [-1.0058699 ,  0.        ,  0.        , ...,  0.3767955 ,
         0.2252788 ,  0.76809537]])

In [63]:
#X3=X.iloc[:,np.r_[0,3:24]] -sliced the summit_elec and Vertical_drop columns
#X3.columns is the respective column names
#X3_test = np.delete(X2_test,[1,2],1) slicing the X2 train

X3_train = np.delete(X2_train,[1,2],1)
X3_test = np.delete(X2_test,[1,2],1)
y3_train = y_train
y3_test = y_test

In [64]:
X3_train.shape

(131, 22)

In [65]:
#all third model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [66]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred3 = model3.predict(X3_test)

In [67]:
from sklearn.metrics import explained_variance_score,mean_absolute_error
evs3 = explained_variance_score(y3_test,y_pred3)
mae3 = mean_absolute_error(y3_test,y_pred3)

In [68]:
evs3

0.6545153484696733

In [69]:
mae3

6.761665050950797

In [70]:
lm3.intercept_

56.40102679996302

In [71]:
lm3.coef_

array([ 7.35778363e-01,  1.22124533e-15, -8.88178420e-16,  1.77635684e-15,
        5.11989356e-01,  1.51983089e+00,  5.05345979e-01,  8.70083449e-01,
       -1.13265830e+00,  6.05676422e-01,  7.56193715e-01,  1.79646257e+00,
        1.53412475e+00, -1.01782589e+00,  1.41775691e+00,  1.69254854e+00,
       -3.62744198e-02, -1.29291021e+00,  9.91288152e+00, -1.83981195e+00,
       -3.82701382e-01, -2.15438346e+00])

In [72]:
lm3.coef_.shape

(22,)

In [73]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: X3=X.iloc[:,np.r_[0,3:24]]
df_coef3 = pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])
df_coef3

Coefficient
Unnamed: 0         7.357784e-01
trams              1.221245e-15
fastEight          8.881784e-16
fastSixes          1.776357e-15
fastQuads          5.119894e-01
quad               1.519831e+00
triple             5.053460e-01
double             8.700834e-01
surface            1.132658e+00
total_chairs       6.056764e-01
Runs               7.561937e-01
TerrainParks       1.796463e+00
LongestRun_mi      1.534125e+00
SkiableTerrain_ac  1.017826e+00
Snow Making_ac     1.417757e+00
daysOpenLastYear   1.692549e+00
yearsOpen          3.627442e-02
averageSnowfall    1.292910e+00
AdultWeekday       9.912882e+00
projectedDaysOpen  1.839812e+00
NightSkiing_ac     3.827014e-01
clusters           2.154383e+00

In [74]:
df_coef3.sort_values(by=['Coefficient'],ascending=False).head(11)

Coefficient
AdultWeekday          9.912882
clusters              2.154383
projectedDaysOpen     1.839812
TerrainParks          1.796463
daysOpenLastYear      1.692549
LongestRun_mi         1.534125
quad                  1.519831
Snow Making_ac        1.417757
averageSnowfall       1.292910
surface               1.132658
SkiableTerrain_ac     1.017826

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.7434902015412724e+21|109791169954.2541  |-|
| Model 2. | 0.6540519043831821|6.923908686959541 |'state'|
| Model 3. | 0.6545153484696733 | 6.761665050950797|'state','summit_elev','base_elev'|

Model Selection:

Model 3 is having the lowest Mean Absolute Error,So Model 3 is preffered